src/control/uliftable.lean

```diff
@@ -0,0 +1,141 @@
+
+/-- Given a universe polymorphic functor `M.{u}`, this class helps move from
+`M.{u}` to `M.{v}` and back -/
+-- class uliftable (f : Type u₀ → Type u₁) (g : Type v₀ → Type v₁) :=
+-- (H : Type.{max u₀ v₀} → Type.{max u₁ v₁})
+-- (congr_f {α} : f α ≃ H (ulift α))
+-- (congr_g {α} : g α ≃ H (ulift.{max v₀ u₀} α))
+
+class uliftable (f : Type u₀ → Type u₁) (g : Type v₀ → Type v₁) :=
+(congr [] {α β} : α ≃ β → f α ≃ g β)
+
+namespace uliftable
+
+/-- The most common practical use `uliftable` (together with `up`), this function takes `x : M.{u} α` and lifts it
+to M.{max u v} (ulift.{v} α) -/
+@[reducible]
+def up {f : Type u₀ → Type u₁} {g : Type (max u₀ v₀) → Type v₁} [uliftable f g]
+  {α} : f α → g (ulift α) :=
+(uliftable.congr f g equiv.ulift.symm).to_fun
+
+/-- The most common practical use of `uliftable` (together with `up`), this function takes `x : M.{max u v} (ulift.{v} α)`
+and lowers it to `M.{u} α` -/
+@[reducible]
+def down {f : Type u₀ → Type u₁} {g : Type (max u₀ v₀) → Type v₁} [uliftable f g]
+  {α} : g (ulift α) → f α :=
+(uliftable.congr f g equiv.ulift.symm).inv_fun
+
+/-- convenient shortcut to avoid manipulating `ulift` -/
+def adapt_up {F : Type u₀ → Type u₁} {G : Type (max u₀ v₀) → Type v₁}
+  [uliftable F G] [monad G] {α β}
+  (x : F α) (f : α → G β) : G β :=
+up x >>= f ∘ ulift.down
+
+/-- convenient shortcut to avoid manipulating `ulift` -/
+def adapt_down {F : Type (max u₀ v₀) → Type u₁} {G : Type v₀ → Type v₁}
+  [L : uliftable G F] [monad F] {α β}
+  (x : F α) (f : α → G β) : G β :=
+@down.{v₀ v₁ (max u₀ v₀)} G F L β $ x >>= @up.{v₀ v₁ (max u₀ v₀)} G F L β ∘ f
+
+def up_map {F : Type u₀ → Type u₁} {G : Type.{max u₀ v₀} → Type v₁} [inst : uliftable F G] [functor G]
+  {α β} (f : α → β) (x : F α) : G β :=
+functor.map (f ∘ ulift.down) (up x)
+
+def down_map {F : Type.{max u₀ v₀} → Type u₁} {G : Type → Type v₁} [inst : uliftable G F] [functor F]
+  {α β} (f : α → β) (x : F α) : G β :=
+down (functor.map (ulift.up ∘ f) x : F (ulift β))
+
+@[simp]
+lemma up_down  {f : Type u₀ → Type u₁} {g : Type (max u₀ v₀) → Type v₁} [uliftable f g]
+  {α} (x : g (ulift α)) : up (down x : f α) = x := (uliftable.congr f g equiv.ulift.symm).right_inv _
+
+@[simp]
+lemma down_up  {f : Type u₀ → Type u₁} {g : Type (max u₀ v₀) → Type v₁} [uliftable f g]
+  {α} (x : f α) : down (up x : g _) = x := (uliftable.congr f g equiv.ulift.symm).left_inv _
+
+end uliftable
+
+open ulift
+
+instance : uliftable id id :=
+{ congr := λ α β F, F }
+
+/-- reduce the equivalence between two state monads to the equivalence between
+their respective function spaces -/
+def state_t.equiv {m₁ : Type u₀ → Type v₀} {m₂ : Type u₁ → Type v₁}
+  {α₁ σ₁ : Type u₀} {α₂ σ₂ : Type u₁} (F : (σ₁ → m₁ (α₁ × σ₁)) ≃ (σ₂ → m₂ (α₂ × σ₂))) :
+  state_t σ₁ m₁ α₁ ≃ state_t σ₂ m₂ α₂ :=
+{ to_fun := λ ⟨f⟩, ⟨F f⟩,
+  inv_fun := λ ⟨f⟩, ⟨F.symm f⟩,
+  left_inv := λ ⟨f⟩, congr_arg state_t.mk $ F.left_inv _,
+  right_inv := λ ⟨f⟩, congr_arg state_t.mk $ F.right_inv _ }
```

Comment:
Perhaps this should be moved to somewhere with generic stuff about `state_t`, as it's not about `uliftable`?